dex/src/main/scala/com/wavesplatform/dex/api/routes/PathMatchers.scala

```diff
@@ -19,16 +19,34 @@ object PathMatchers {
 
   }
 
-  val AssetPairPM: PathMatcher1[AssetPair] = AkkaMatchers.Segments(2).flatMap {
-    case a1 :: a2 :: Nil => AssetPair.createAssetPair(a1, a2).toOption
-    case _ => None
+  val AssetPairPM: PathMatcher1[Either[InvalidAsset, AssetPair]] = AkkaMatchers.Segments(2).flatMap {
+    case a1 :: a2 :: Nil =>
+      Option(try Right(AssetPair.createAssetPair(a1, a2).get)
+      catch {
+        case e: Exception =>
+          if (e.getMessage.contains(a1)) Left(InvalidAsset(a1, e.getMessage))
+          else Left(InvalidAsset(a2, e.getMessage))
+      })
+    case _ => Option(Left(InvalidAsset(null, "Unexpected error")))
   }
 
-  val AssetPM: PathMatcher1[Asset] = AkkaMatchers.Segment.flatMap { s =>
-    AssetPair.extractAsset(s).toOption
-  }
-
-  object ByteStrPM extends Base58[ByteStr](ByteStr.decodeBase58(_).toOption)
+  object AssetPM
+      extends Base58[Either[InvalidAsset, Asset]](s =>
+        Option(try Right(AssetPair.extractAsset(s).get)
```

Comment:
`AssetPair.extractAsset(s)` returns `Try`. Why do you extract it and catch again? :) Note, there is `toEither` method in `Try`

Reply:
Fixed